### PR TITLE
Adding crud and unit of work

### PR DESCRIPTION
**Pull request by AntonBakum**

### CODE REVIEW

File: pet-api/pet-api/Domain/Entities/Pagination/ProductParameters.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿namespace pet_api.Domain.Entities.Pagination
+{
+    public class ProductParameters
+    {
+        const int maxPageSize = 50;
```
Comment (ValentynCGL) [inline pet-api/pet-api/Domain/Entities/Pagination/ProductParameters.cs:5]: use private const. const should be named from a big letter

File: pet-api/pet-api/Controllers/CategoriesController.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            Category category = await _unitOfWork.CategoryRepository.GetById(id);
+            if (category == null)
+            {
+                return NotFound();
+            }
+            return Ok(category);
+        }
+
+        [HttpGet]
+        public async Task<ActionResult> GetCategories()
```
Comment (ValentynCGL) [inline pet-api/pet-api/Controllers/CategoriesController.cs]: add type for the ActionResult`<Type here>`

File: pet-api/pet-api/Controllers/ProductsController.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            Product product = await _unitOfWork.ProductRepository.GetById(id);
+            if (product == null)
+            {
+                return NotFound();
+            }
+            return Ok(product);
+        }
+
+        [HttpGet]
+        public async Task<ActionResult> GetProducts([FromQuery] ProductParameters productParameters)
```
Comment (ValentynCGL) [inline pet-api/pet-api/Controllers/ProductsController.cs]: add type for ActionResult

File: pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
+﻿using Dapper;
+using pet_api.Domain.Entities;
+using pet_api.Domain.Entities.Pagination;
+using pet_api.Domain.Interfaces;
+using System.Data;
+
```
Comment (ValentynCGL) [inline pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs:9]: Need to add some error handling like try-catch

File: pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        public class ProductRepository : IRepository<Product>
+        {
+            private readonly IDatabaseContext _context;
+
+            public ProductRepository(IDatabaseContext context)
+            {
+                _context = context;
+            }
+
+            public async Task Create(Product entity)
```
Comment (YevheniiDomnich) [inline pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs]: Where is this parameter used?

File: pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                    await connection.ExecuteAsync(sqlQuery, parameters);
+                }
+            }
+            public async Task<PagedList<Product>> GetByPage(ProductParameters productParameters)
+            {
+                string sqlQuery = "SELECT * FROM Categories";
+                IEnumerable<Product> products;
+                using (var connection = _context.SqlConnection)
+                {
+                    products = await connection.QueryAsync<Product>(sqlQuery);
```
Comment (YevheniiDomnich) [inline pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs]: You are paginating in memory. This is bad. What is you have to retrieve 1 million records from the DB? The memory is limited. You should paginate it in the DB

File: pet-api/pet-api/Domain/Entities/Pagination/ProductParameters.cs (hunk 1 above)
Comment (AntonBakum) [reply pet-api/pet-api/Domain/Entities/Pagination/ProductParameters.cs:5]: private modifier has been added

File: pet-api/pet-api/Controllers/CategoriesController.cs (hunk 2 above)
Comment (AntonBakum) [reply pet-api/pet-api/Controllers/CategoriesController.cs]: ActionResult<IEnumerable <Products>>

File: pet-api/pet-api/Controllers/ProductsController.cs (hunk 3 above)
Comment (AntonBakum) [reply pet-api/pet-api/Controllers/ProductsController.cs]: ActionResult contains IEnumerable<Product> parameter

File: pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs (hunk 4 above)
Comment (AntonBakum) [reply pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs:9]: I added an error handling into my UnitOfWork. Also errors are writed to the log file, it is very helpful feature

File: pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs (hunk 5 above)
Comment (AntonBakum) [reply pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs]: This logic has been changed

File: pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs (hunk 6 above)
Comment (AntonBakum) [reply pet-api/pet-api/Infrastructure/DAL/Repositories/ProductRepository.cs]: Paginating using SQL queries has been added

File: pet-api/pet-api/Controllers/CategoriesController.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            }
+            return NotFound();
+        }
+
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<Category>>> GetCategories()
+        {
+            await _unitOfWork.BeginTransaction();
+            IEnumerable<Category> categories = await _unitOfWork.CategoryRepository.GetAll();
+            await _unitOfWork.Commit();
```
Comment (YevheniiDomnich) [inline pet-api/pet-api/Controllers/CategoriesController.cs]: what are you committing here?

File: pet-api/pet-api/Controllers/ProductsController.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            }
+            return Ok(product);
+        }
+
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<Product>>> GetProducts([FromQuery] ProductParameters productParameters)
+        {
+            await _unitOfWork.BeginTransaction();
+            IEnumerable<Product> products = await _unitOfWork.ProductRepository.GetByPage(productParameters);
+            await _unitOfWork.Commit();
```
Comment (YevheniiDomnich) [inline pet-api/pet-api/Controllers/ProductsController.cs]: do we need it here?

File: pet-api/pet-api/Infrastructure/DAL/UnitOfWork.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                }
+                disposedValue = true;
+            }
+        }
+
+        /*~UnitOfWork()
```
Comment (YevheniiDomnich) [inline pet-api/pet-api/Infrastructure/DAL/UnitOfWork.cs]: delete comment

**Review by YevheniiDomnich — APPROVED** [review]

**Review by ValentynCGL — APPROVED** [review]

**Pull request merged**